src/hdbscan.jl

```diff
@@ -0,0 +1,296 @@
+hasstability(c::HdbscanCluster) = c.stability != 0
+function compute_stability(c::HdbscanCluster, λbirth)
+    c.stability += sum(c.λp.-λbirth)
+end
+
+"""
+    HdbscanResult(k, minpts, clusters)
+- `k`: we will define "core distance of point A" as the distance between point A and the `k` th neighbor point of point A.
+- `min_cluster_size`: minimum number of points in the cluster
+- `clusters`: result vector of clusters
+"""
+mutable struct HdbscanResult
+    k::Int
+    min_cluster_size::Int
+    clusters::Vector{HdbscanCluster}
+end
+
+"""
+    hdbscan(points::AbstractMatrix, k::Int, min_cluster_size::Int; gen_mst::Bool=true, mst=nothing)
+Density-Based Clustering Based on Hierarchical Density Estimates.
+This algorithm performs clustering as follows.
+1. generate a minimum spanning tree
+2. build a HDBSCAN hierarchy
+3. extract the target cluster
+4. generate the list of cluster assignment for each point
+The detail is so complex it is difficult to explain the detail in here. But, if you want to know more about this algorithm, you should read [this docs](https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html).
+
+# Parameters
+- `points`: the d×n matrix, where each column is a d-dimensional coordinate of a point
+- `k`: we will define "core distance of point A" as the distance between point A and the `k` th neighbor point of point A.
+- `min_cluster_size`: minimum number of points in the cluster
+- `gen_mst`: whether to generate minimum-spannig-tree or not
+- `mst`: when is specified and `gen_mst` is false, new mst won't be generated
+"""
+function hdbscan(points::AbstractMatrix, k::Int, min_cluster_size::Int; gen_mst::Bool=true, mst=nothing)
+    if min_cluster_size < 1
+        throw(DomainError(min_cluster_size, "The `min_cluster_size` must be greater than or equal to 1"))
+    end
+    n = size(points, 1)
+    if gen_mst
+        #calculate core distances for each point
+        core_dists = core_dist(points, k)
+        #calculate mutual reachability distance between any two points
+        mrd = mutual_reachability_distance(core_dists, points)
+        #compute a minimum spanning tree by prim method
+        mst = prim(mrd, n)
+    elseif mst == nothing
+        throw(ArgumentError("if you set `gen_mst` to false, you must pass a minimum spanning tree as `mst`"))
+    end
+    #build a HDBSCAN hierarchy
+    hierarchy = build_hierarchy(mst, min_cluster_size)
+    #extract the target cluster
+    extract_cluster!(hierarchy)
+    #generate the list of cluster assignment for each point
+    result = HdbscanCluster[]
+    noise_points = fill(-1, n)
+    for (i, j) in enumerate(hierarchy[2n-1].children)
+        c = hierarchy[j]
+        push!(result, c)
+        for k in c.points
+            noise_points[k] = 0
+        end
+    end
+    push!(result, HdbscanCluster(true, Int[]))
+    result[end].points = findall(x->x==-1, noise_points)
+    return HdbscanResult(k, min_cluster_size, result)
+end
+
+function core_dist(points, k)
```

Comment:
@alyst 
I changed code with Distances.jl like this:
`dists = vec(pairwise(SqEuclidean(), points[i:i, :], points; dims=1))`
But it seems to be harmful. Look at the following benchmark:
```
julia> @benchmark core_distances(points, 5)
BenchmarkTools.Trial: 100 samples with 1 evaluation.
 Range (min … max):  49.296 ms … 64.033 ms  ┊ GC (min … max): 2.63% … 2.00%
 Time  (median):     49.839 ms              ┊ GC (median):    3.05%
 Time  (mean ± σ):   50.050 ms ±  1.493 ms  ┊ GC (mean ± σ):  2.96% ± 0.35%

     ▅▂█▃▇▂ ▃▃                                                 
  ▃▄██████████▅▃▅▁▁▁▃▃▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃ ▃
  49.3 ms         Histogram: frequency by time        53.7 ms <

 Memory estimate: 198.36 MiB, allocs estimate: 5501.

julia> @benchmark core_dist(points, 5)
BenchmarkTools.Trial: 161 samples with 1 evaluation.
 Range (min … max):  29.093 ms … 49.369 ms  ┊ GC (min … max): 4.43% … 6.92%
 Time  (median):     29.666 ms              ┊ GC (median):    5.43%
 Time  (mean ± σ):   31.224 ms ±  3.439 ms  ┊ GC (mean ± σ):  5.11% ± 0.72%

  ▃█▃▂                                                         
  ████▄▂▃▂▂▁▃▁▁▂▃▂▁▃▁▁▁▁▁▁▁▂▁▂▄▂▂▃▃▃▃▂▁▁▃▁▃▁▁▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▂ ▂
  29.1 ms         Histogram: frequency by time        42.4 ms <

 Memory estimate: 195.76 MiB, allocs estimate: 7001.

```
`points` is a 500x100 random matrix. I don't know well about optimization, so I wonder why the performance got worse. which should I use the new code or old code?

Reply:
What I have originally meant is to call `dists = pairwise(Euclidean(), points; dims=2)` once to calculate the distances between all pairs of points (see the [Distances.jl readme](https://github.com/JuliaStats/Distances.jl?tab=readme-ov-file#computing-pairwise-distances)).
I think this have to be done outside the loop, and maybe even before you call `core_dists()`, because you also need the distances in `hdbscan_graph()`.
Then inside the loop, you just take iterate over the columns of dists (e.g. `for (i, i_dists) in enumeate(eachcol(dists))`) to select the k smallest distances.

Note that if you do it once, and only pass `dists` matrix to `core_distances()` and `hdbscan_graph()`, you can allow the user to specify any other metric as a parameter to `hdbscan(...; metric = SqEuclidean())` (and then you call `dists = pairwise(metric, points, dims=2)`.